pkg/kv/kvserver/replica_raftstorage.go

```diff
@@ -57,26 +57,69 @@ var snapshotIngestAsWriteThreshold = settings.RegisterByteSizeSetting(
 	}())
 
 // replicaRaftStorage implements the raft.Storage interface.
+//
+// All mutating calls to raft.RawNode require that r.mu is held. All read-only
+// calls to raft.RawNode require that r.mu is held at least for reads.
+//
+// All methods implementing raft.Storage are called from within, or on behalf of
+// a RawNode. When called from within RawNode, r.mu is held necessarily (and
+// maybe r.raftMu). When called from outside RawNode (on behalf of a RawNode
+// "snapshot"), the caller must hold r.raftMu and/or r.mu.
+//
+// RawNode has the in-memory "unstable" state which services most of its needs.
+// Most RawNode.Step updates are completed in memory, while only holding r.mu.
+//
+// RawNode falls back to reading from Storage when it does not have the needed
+// state in memory. For example, the leader may need to read log entries from
+// storage to construct a log append requests for a follower, or a follower may
+// need to interact with its storage upon receiving such a request to check
+// whether the appended log slice is consistent with raft rules.
+//
+// (1) There is a guarantee that everything RawNode reads from Storage has no
+// in-flight writes. Raft always reads state that it knows to be stable (meaning
+// it does not have pending writes) and, in some cases, also synced / durable.
+// Storage acknowledges completed writes / syncs back to RawNode, under r.mu.
+//
+// This guarantee explains why holding only r.mu is sufficient for calling
+// storage methods from within or on behalf of RawNode - it ensures that RawNode
+// remains consistent. However, r.mu is a widely used mutex, and not recommended
+// for IO. There is a second guarantee that we can rely on.
+//
+// (2) Stable state in raft.Storage is always mutated while holding r.raftMu,
+// which is an un-contended "IO" mutex and is allowed to be held longer. Most
+// writes are extracted from RawNode while holding r.mu (in the Ready() loop),
+// and handed over to storage under r.raftMu. There are a few cases when CRDB
+// synthesizes the writes (e.g. during a range split / merge) under r.raftMu.
```

Comment:
need to check whether the synthesized writes hold only `raftMu` or both (both I think?)